File: src/db/sessions.ts
/**
 * Session journal queries + the persist transaction that writes a structured
 * session and folds its techniques into the library (plan §5.6).
 */

import type { SQLiteDatabase } from 'expo-sqlite';

import { upsertTechnique } from './techniques';
import type {
  Round,
  Session,
  SessionRow,
  SessionTechnique,
  StructuredSession,
} from './types';

function parseArray(json: string | null): string[] {
  if (!json) return [];
  try {
    const v = JSON.parse(json);
    return Array.isArray(v) ? v.filter((x) => typeof x === 'string') : [];
  } catch {
    return [];
  }
}

function parseRounds(json: string | null): Round[] {
  if (!json) return [];
  try {
    const v = JSON.parse(json);
    return Array.isArray(v) ? v : [];
  } catch {
    return [];
  }
}

function mapSession(row: SessionRow): Session {
  return {
    id: row.id,
    createdAt: row.created_at,
    rawTranscript: row.raw_transcript,
    summary: row.summary,
    wentWell: parseArray(row.went_well),
    toImprove: parseArray(row.to_improve),
    rounds: parseRounds(row.rounds),
    tags: parseArray(row.tags),
  };
}

/** Reverse-chronological list for the journal. */
export async function listSessions(db: SQLiteDatabase): Promise<Session[]> {
  const rows = await db.getAllAsync<SessionRow>(
    'SELECT * FROM sessions ORDER BY created_at DESC'
  );
  return rows.map(mapSession);
}

export async function getSession(
  db: SQLiteDatabase,
  id: number
): Promise<Session | null> {
  const row = await db.getFirstAsync<SessionRow>(
    'SELECT * FROM sessions WHERE id = ?',
    [id]
  );
  return row ? mapSession(row) : null;
}

/** Techniques linked to a session, with the per-session notes. */
export async function getSessionTechniques(
  db: SQLiteDatabase,
  sessionId: number
): Promise<SessionTechnique[]> {
  return db.getAllAsync<SessionTechnique>(
    `SELECT t.id AS techniqueId, t.name AS name, t.category AS category,
            t.position AS position, st.notes AS notes
     FROM session_techniques st
     JOIN techniques t ON t.id = st.technique_id
     WHERE st.session_id = ?
     ORDER BY t.name COLLATE NOCASE`,
    [sessionId]
  );
}

/** Update the user-editable, self-assessment fields of a session. */
export async function updateSession(
  db: SQLiteDatabase,
  id: number,
  fields: { summary: string | null; wentWell: string[]; toImprove: string[]; tags: string[] }
): Promise<void> {
  await db.runAsync(
    `UPDATE sessions
     SET summary = ?, went_well = ?, to_improve = ?, tags = ?
     WHERE id = ?`,
    [
      fields.summary,
      JSON.stringify(fields.wentWell),
      JSON.stringify(fields.toImprove),
      JSON.stringify(fields.tags),
      id,
    ]
  );
}

export async function deleteSession(db: SQLiteDatabase, id: number): Promise<void> {
  // session_techniques rows cascade; technique library rows are intentionally kept.
  await db.runAsync('DELETE FROM sessions WHERE id = ?', [id]);
}

/**
 * Persist a structured session + its techniques in one transaction.
 * Returns the new session id. Steps:
 *   1. insert the sessions row (JSON columns stringified)
 *   2. upsert each technique (dedup by normalized name)
 *   3. insert session_techniques join rows with per-session notes
 */
export async function persistSession(
  db: SQLiteDatabase,
  args: { rawTranscript: string; structured: StructuredSession; createdAt?: string }
): Promise<number> {
  const now = args.createdAt ?? new Date().toISOString();
  const s = args.structured;

  let sessionId = 0;
  await db.withTransactionAsync(async () => {
    const res = await db.runAsync(
      `INSERT INTO sessions
         (created_at, raw_transcript, summary, went_well, to_improve, rounds, tags)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
      [
        now,
        args.rawTranscript,
        s.summary ?? null,
        JSON.stringify(s.went_well ?? []),
        JSON.stringify(s.to_improve ?? []),
        JSON.stringify(s.rounds ?? []),
        JSON.stringify(s.tags ?? []),
      ]
    );
    sessionId = res.lastInsertRowId;

    // Dedup techniques within this single note so we never violate the join PK.
    const seen = new Map<number, string>();
    for (const tech of s.techniques ?? []) {
      if (!tech.name?.trim()) continue;
      const techniqueId = await upsertTechnique(db, {
        name: tech.name,
        category: tech.category ?? null,
        position: tech.position ?? null,
        description: tech.session_notes ?? null,
        now,
      });
      if (seen.has(techniqueId)) continue; // same technique mentioned twice in one note
      seen.set(techniqueId, tech.session_notes ?? '');
      await db.runAsync(
        'INSERT INTO session_techniques (session_id, technique_id, notes) VALUES (?, ?, ?)',
        [sessionId, techniqueId, tech.session_notes ?? null]
      );
    }
  });

  return sessionId;
}

/** Total counts for empty-state / settings display. */
export async function getCounts(
  db: SQLiteDatabase
): Promise<{ sessions: number; techniques: number }> {
  const s = await db.getFirstAsync<{ n: number }>('SELECT COUNT(*) AS n FROM sessions');
  const t = await db.getFirstAsync<{ n: number }>('SELECT COUNT(*) AS n FROM techniques');
  return { sessions: s?.n ?? 0, techniques: t?.n ?? 0 };
}

/** Wipe all data (settings "clear data"). */
export async function clearAllData(db: SQLiteDatabase): Promise<void> {
  await db.withTransactionAsync(async () => {
    await db.execAsync(
      'DELETE FROM session_techniques; DELETE FROM sessions; DELETE FROM techniques;'
    );
  });
}

File: server/app/repositories/sessions.py
"""Session journal queries and the persist transaction.

JSON-array columns (`went_well`, `to_improve`, `rounds`, `tags`) are stored as
TEXT and decoded here, mirroring the React Native app's `db/sessions.ts`.
"""

from __future__ import annotations

import json
import sqlite3
from datetime import datetime, timezone

from ..models import (
    Round,
    Session,
    SessionListItem,
    SessionTechnique,
    SessionUpdate,
    StructuredSession,
)
from ..text import derive_title, normalize_name, to_title
from .techniques import upsert_technique


def _parse_str_list(raw: str | None) -> list[str]:
    if not raw:
        return []
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return []
    return [item for item in value if isinstance(item, str)] if isinstance(value, list) else []


def _parse_rounds(raw: str | None) -> list[Round]:
    if not raw:
        return []
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return []
    if not isinstance(value, list):
        return []
    return [Round.model_validate(item) for item in value if isinstance(item, dict)]


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def list_sessions(conn: sqlite3.Connection) -> list[SessionListItem]:
    """Reverse-chronological list for the journal."""
    rows = conn.execute(
        """
        SELECT id, created_at, title, summary, tags, raw_transcript
          FROM sessions
         ORDER BY created_at DESC
        """
    ).fetchall()
    return [
        SessionListItem(
            id=row["id"],
            created_at=row["created_at"],
            title=derive_title(row["title"], row["summary"], row["raw_transcript"]),
            summary=row["summary"],
            tags=_parse_str_list(row["tags"]),
        )
        for row in rows
    ]


def get_session(conn: sqlite3.Connection, session_id: int) -> Session | None:
    row = conn.execute("SELECT * FROM sessions WHERE id = ?", (session_id,)).fetchone()
    if row is None:
        return None

    return Session(
        id=row["id"],
        created_at=row["created_at"],
        raw_transcript=row["raw_transcript"],
        title=derive_title(row["title"], row["summary"], row["raw_transcript"]),
        summary=row["summary"],
        went_well=_parse_str_list(row["went_well"]),
        to_improve=_parse_str_list(row["to_improve"]),
        rounds=_parse_rounds(row["rounds"]),
        tags=_parse_str_list(row["tags"]),
        techniques=get_session_techniques(conn, session_id),
    )


def get_session_techniques(
    conn: sqlite3.Connection, session_id: int
) -> list[SessionTechnique]:
    rows = conn.execute(
        """
        SELECT t.id AS technique_id, t.name, t.category, t.position, st.notes
          FROM session_techniques st
          JOIN techniques t ON t.id = st.technique_id
         WHERE st.session_id = ?
         ORDER BY t.name COLLATE NOCASE
        """,
        (session_id,),
    ).fetchall()
    return [SessionTechnique.model_validate(dict(row)) for row in rows]


def update_session(
    conn: sqlite3.Connection, session_id: int, fields: SessionUpdate
) -> bool:
    """Update the user-editable fields. Returns False if the session is missing."""
    cursor = conn.execute(
        """
        UPDATE sessions
           SET title = ?, summary = ?, went_well = ?, to_improve = ?, tags = ?
         WHERE id = ?
        """,
        (
            to_title(fields.title) if fields.title else None,
            fields.summary,
            json.dumps(fields.went_well),
            json.dumps(fields.to_improve),
            json.dumps(fields.tags),
            session_id,
        ),
    )
    return cursor.rowcount > 0


def delete_session(conn: sqlite3.Connection, session_id: int) -> bool:
    """Delete a session and the library techniques it brought in.

    `session_techniques` rows cascade. A technique may be linked to several
    sessions, so only the ones left with no remaining session are removed;
    survivors have their counters recomputed from the rows that are actually
    left, which is self-healing in a way `times_trained - 1` would not be.
    """
    linked = [
        row["technique_id"]
        for row in conn.execute(
            "SELECT technique_id FROM session_techniques WHERE session_id = ?",
            (session_id,),
        ).fetchall()
    ]

    cursor = conn.execute("DELETE FROM sessions WHERE id = ?", (session_id,))
    if cursor.rowcount == 0:
        return False

    for technique_id in linked:
        remaining = conn.execute(
            "SELECT COUNT(*) AS n FROM session_techniques WHERE technique_id = ?",
            (technique_id,),
        ).fetchone()["n"]

        if remaining == 0:
            conn.execute("DELETE FROM techniques WHERE id = ?", (technique_id,))
            continue

        conn.execute(
            """
            UPDATE techniques
               SET times_trained = (
                     SELECT COUNT(*) FROM session_techniques
                      WHERE technique_id = techniques.id
                   ),
                   first_seen = COALESCE((
                     SELECT MIN(s.created_at) FROM session_techniques st
                       JOIN sessions s ON s.id = st.session_id
                      WHERE st.technique_id = techniques.id
                   ), first_seen),
                   last_seen = COALESCE((
                     SELECT MAX(s.created_at) FROM session_techniques st
                       JOIN sessions s ON s.id = st.session_id
                      WHERE st.technique_id = techniques.id
                   ), last_seen)
             WHERE id = ?
            """,
            (technique_id,),
        )

    return True


def persist_session(
    conn: sqlite3.Connection,
    *,
    raw_transcript: str,
    structured: StructuredSession,
    created_at: str | None = None,
) -> int:
    """Write a structured session and fold its techniques into the library.

    Returns the new session id. The caller's connection provides the transaction.
    """
    now = created_at or _now()

    cursor = conn.execute(
        """
        INSERT INTO sessions
            (created_at, raw_transcript, title, summary,
             went_well, to_improve, rounds, tags)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """,
        (
            now,
            raw_transcript,
            structured.title.strip() or None,
            structured.summary or None,
            json.dumps(structured.went_well),
            json.dumps(structured.to_improve),
            json.dumps([r.model_dump() for r in structured.rounds]),
            json.dumps(structured.tags),
        ),
    )
    session_id = int(cursor.lastrowid)

    # Dedup within this single note, so we neither violate the join table's PK
    # nor double-count `times_trained`. The check must come BEFORE the upsert:
    # upserting first would bump the counter for each mention even though only
    # one join row is written.
    seen: set[str] = set()
    for technique in structured.techniques:
        if not technique.name.strip():
            continue

        name_norm = normalize_name(technique.name)
        if name_norm in seen:
            continue  # same technique mentioned twice in one debrief
        seen.add(name_norm)

        technique_id = upsert_technique(
            conn,
            name=technique.name,
            category=technique.category or None,
            position=technique.position,
            description=technique.session_notes or None,
            now=now,
        )

        conn.execute(
            "INSERT INTO session_techniques (session_id, technique_id, notes)"
            " VALUES (?, ?, ?)",
            (session_id, technique_id, technique.session_notes or None),
        )

    return session_id
